fdbserver/DataDistribution.actor.cpp

```diff
@@ -1056,104 +1421,749 @@ struct DDTeamCollection {
-				vector<UID> processIDs;
+			//Step 5: Restore machine from its representative process team and get the machine team
+			if ( bestTeam.size() == configuration.storageTeamSize ) {
+				vector<Standalone<StringRef>> machineIDs;
+				for ( auto process = bestTeam.begin(); process < bestTeam.end(); process++ ) {
+					Standalone<StringRef> machine_id  = server_info[**process]->lastKnownInterface.locality.machineId().get();
+					assert( machine_info.find(machine_id) != machine_info.end() );
+					machineIDs.push_back(machine_id);
+				}
+
+				std::sort(machineIDs.begin(), machineIDs.end());
+				if( !machineTeamExists(machineIDs) ) {
+					addMachineTeam(machineIDs.begin(), machineIDs.end());
+					addedMachineTeams++;
+				}
+			} else {
+				TraceEvent(SevWarn, "DataDistributionBuildTeams", masterId).detail("Primary", primary)
+					.detail("Reason","Unable to make desired machine Teams");
+				break;
+			}
+
+			if ( ++loopCount > 2*machineTeamsToBuild*(configuration.storageTeamSize+1) ) { // MX: Q: why is this number?
+				break;
+			}
+		}
+
+		return addedMachineTeams;
+	}
+
+	bool isMachineTeamHealthy( vector<Standalone<StringRef>> const &machineIDs ) {
+		int healthyNum = 0;
+
+		// A healthy machine team should have the desired number of machines
+		if ( machineIDs.size() != configuration.storageTeamSize )
+			return false;
+
+		for ( auto &id : machineIDs ) {
+			auto &machine = machine_info[id];
+			if ( isMachineHealthy(machine) ) {
+				healthyNum++;
+			}
+		}
+		if ( healthyNum == machineIDs.size() )
+			return true;
+		else
+			return false;
+	}
+
+	bool isMachineTeamHealthy( Reference<TCMachineTeamInfo> const &machineTeam ) {
+		int healthyNum = 0;
+
+		// A healthy machine team should have the desired number of machines
+		if ( machineTeam->size() != configuration.storageTeamSize )
+			return false;
+
+		for ( auto &machine : machineTeam->machines ) {
+			if ( isMachineHealthy(machine) ) {
+				healthyNum++;
+			}
+		}
+		if ( healthyNum == machineTeam->machines.size() )
+			return true;
+		else
+			return false;
+	}
+
+	bool isMachineHealthy( Reference<TCMachineInfo> const &machine ) {
+		if ( !machine.isValid() ||machine_info.find(machine->machineID) == machine_info.end()
+			 || machine_info[machine->machineID]->serversOnMachine.empty() ) {
+			//Debug trace
+			if ( !machine.isValid() )
+				TraceEvent(SevWarn, "InvalidMachineTeam").detail("IsValid", machine.isValid());
+			else
+				TraceEvent(SevWarn, "InvalidMachineTeam").detail("IsValid", machine.isValid())
+						.detail("InMachineInfo", machine_info.find(machine->machineID) != machine_info.end())
+						.detail("ServerNumber", machine_info[machine->machineID]->serversOnMachine.size());
+
+			return false;
+		}
+
+		bool healthy = false;
+		// Healthy machine has at least one healthy server
+		for ( auto &server : machine->serversOnMachine ) {
+			if ( !server_status.get(server->id).isUnhealthy() ) {
+				healthy = true;
+				break;
+			}
+		}
```

Comment:
Not required to change, but take a look at the contents of <algorithms> sometime.  There's generic functions in there that can save from writing boilerplate code/loops.

For example, this could be with something like

    Servers& onMachine = machine->serversOnMachine;
    return std::any_of( onMachine.begin(), onMachine.end(),
            [&](const Machine& m) { return m.get(server->id).isUnhealthy(); } );